---
title: "DNA methylation analysis of SMC phase"
author: "Kennedy Zinn"
date: "2025-02-07"
output: html_document
toc: TRUE
---

```{r, echo = FALSE}
suppressPackageStartupMessages(library(minfi))
suppressPackageStartupMessages(library(limma))
suppressPackageStartupMessages(library(DMRcate))
suppressPackageStartupMessages(library(maxprobes))
suppressPackageStartupMessages(library(readxl))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(missMethyl))
suppressPackageStartupMessages(library(MethylToSNP))
```

# Smooth Muscle Cell (SMC) Comparisons

```{r}
# define path to files

file_folder = "SMC"
```

```{r}
base_sheet = "SampleSheet_X.xlsx"
sheet = gsub("_X", paste0("_", file_folder), base_sheet)

base_path = "/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/dna methyl"

qc_report_name = paste0(file_folder, "qcReport.pdf")
output = paste(file_folder, "Analysis")
  
file_path = file.path(base_path, file_folder)
output_path = file.path(base_path, output)
sheet_path = file.path(base_path, file_folder, sheet)
qc_report_path = file.path(base_path, file_folder, qc_report_name)
```

## Load Data

```{r}
# load the idat files
rgSet <- read.metharray.exp(file_path, force = T)
```

```{r}
# load sample sheet
sample_sheet <- read_excel(sheet_path)
```

## Quality check

```{r}
# Generate QC report
qcReport(rgSet, pdf = qc_report_path, sampNames = sample_sheet$`Sample Name`, sampGroups = sample_sheet$Status)
```

## Pre-Process

```{r}
source("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/dna methyl/process_methyl_microarray.R")
gmSet <- process_methyl_microarray(rgSet)
```

Beta values, which range between [0,1], represent the proportion of methylated DNA copies at a specific CpG site compared to all of the DNA copies at that site. Beta values for the ith CpG site are mathematically defined as,

$$β_i=max(y_i(methyl),0)/(max(y_i(unmethyl),0)+max(y_i(methyl),0)+α),$$

where yi methyl and unmethyl are the intensities measured by the ith methylated and unmethylated probes, respectively. 0 ensures that negative signals that take place after background adjustment are not considered. Alpha, a constant (usually 100), serves as an offset to regularize beta when both methylated and unmethylated instensities are low. Beta values are intuitive, and hence useful for visual presentations of the reads. (Du et al., 2010)

```{r, echo = FALSE}
# Compute beta values
beta_values <- getBeta(gmSet)
colnames(beta_values) <- sample_sheet$`Sample Name`
```

M values are log transformed beta values. Negative M values indicate higher levels of unmethylated DNA copies than methylated DNA copies. Postive M values indicate higher levels of methylated DNA copies than unmethylated. M values close to zero indicate similar intensities between methylated and unmethylated probes. M values are mathematically expressed as,

$$M_i=log2{(max{y(i),methyl}+α)/(max{y(i),unmethyl}+α)}$$

where alpha is a constant to offset dramatic changes resulting from small estimation errors. M values can be expressed in terms of beta values,

$$M_i=log2(β_i/(1 - β_i))$$

and therefore, beta values can be expressed in terms of M values,

$$β_i=(2^Mi)/(2^Mi+1)$$

M values are more statistically sound than beta values.

```{r, echo = FALSE}
Mval <- getM(gmSet)
dim(Mval)
```

It appears that the group with Sentrix ID 204081560070 clusters, suggesting a batch effect.

```{r, echo = FALSE}
# PCA plot of M values
par(mfrow=c(1,1))
plotMDS(Mval, labels=sample_sheet$`Sample Name`, col=as.integer(factor(sample_sheet$Status)))
legend("top",legend=c("Control","Patient"),pch=16,cex=.5,col=1:2)
```

## Differential Methylation Analysis

Design matrix

```{r, echo = FALSE}
batch <- factor(ifelse(sample_sheet$`Sentrix ID` == 204081560070, "batch.effect", "no.batch"))
disease <- factor(sample_sheet$Status)
gender <- factor(sample_sheet$Gender)
design <- model.matrix(~disease + batch + gender, data = sample_sheet)
colnames(design) <- c(levels(disease),levels(batch)[-1])
design
```

## Removing Unwanted Variation with SVA package

Fit model using M values as M values are more statistically valid, beta is more intuitive (Du et al., 2010)

```{r}
# create full and null model
full <- design
null <- model.matrix(~1 + batch, data = sample_sheet)

# applying sva function
library(sva)
n.sv <- num.sv(Mval,full,method = "leek") # estimate number of latent factors
n.sv
# estimate the surrogate variables (variables which are intended to represent the unknown variation)
svobj <- sva(Mval, full, null, n.sv = 3)

# adjust surrogate variable using f-test p values of each cpg site
f_p_val <- f.pvalue(Mval,full,null)
q_val <- p.adjust(f_p_val, method = "BH")

fullSv <- cbind(full, svobj$sv)
nullSv <- cbind(null, svobj$sv)
p_val_sv <- f.pvalue(Mval, fullSv, nullSv)
q_val_sv <- p.adjust(p_val_sv, method = "BH")

# run model while adjusting for surrogate variables
fit <- lmFit(Mval, fullSv)
colnames(fullSv) <- make.names(colnames(fullSv))
contrast <- makeContrasts(patient - control, levels = fullSv)
fitContrast <- contrasts.fit(fit, contrast)
fit <- eBayes(fitContrast)
top <- topTable(fit, adjust = "BH", number = "Inf")
head(top)
```

Visualize results using beta values

```{r, echo = FALSE}
cpgs <- rownames(top)
par(mfrow = c(2,2), cex = 0.5)
for(i in 1:4){
stripchart(beta_values[rownames(beta_values) %in% cpgs[i],]~design[,2],method="jitter",
group.names=c("Control","Patient"), pch = 16,cex=1.5,col=c(4,2), ylab="Beta values",
vertical=TRUE,cex.axis=1.5,cex.lab=1.5)
title(cpgs[i],cex.main=1.5)
}
```

## Differential Variability Analysis

fit model testing variances between controls and patients

```{r, echo = FALSE}
fitvar <- varFit(Mval, design = design, coef = 1:3)
summary(decideTests(fitvar))
topVar <- topVar(fitvar, coef=3)
topVar
```

Visualize top 4 most differentially variable CpGs

```{r, echo = FALSE}
cpgsDV <- rownames(topVar)
par(mfrow=c(2,2), cex = 0.5)
for(i in 1:4){
stripchart(beta_values[rownames(beta_values)==cpgsDV[i],]~design[,2], method="jitter",
group.names=c("Control","Patient"),pch=16,cex=1.5,col=c(4,2),ylab="Beta values",
vertical=TRUE,cex.axis=1.5,cex.lab=1.5)
title(cpgsDV[i],cex.main=1.5)
}
```

Identify the most variable patient sample

```{r, echo = FALSE}
par(mfrow=c(2,2), cex = 0.5)
for(i in 1:4){
stripchart(beta_values[rownames(beta_values)==cpgsDV[i],]~colnames(beta_values), method="jitter",pch=16,cex=1.5,ylab="Beta values",
vertical=TRUE,cex.axis=1.5,cex.lab=1.5)
title(cpgsDV[i],cex.main=1.5)
}
```

## Gene Ontology Region Level Analysis

```{r}
source("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/dna methyl/region_analysis.R")
results.ranges <- region_analysis("patient - control", "0.25", "0.01")
anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b5.hg38)
```

```{r}
gst.region.go <- goregion(results.ranges, all.cpg=rownames(top), 
                       collection= "GO", array.type="EPIC", plot.bias = T, anno = anno, sig.genes = T)
topGSA(gst.region.go[, -ncol(gst.region.go)])
```

We can also test according to a specified gene set using "gsaregion"

## KEGG Region Level Analysis

```{r}
gst.region.kegg <- goregion(results.ranges, all.cpg=rownames(top), 
                       collection= "KEGG", array.type="EPIC", plot.bias = T, anno = anno, sig.genes = T)
topGSA(gst.region.kegg[, -ncol(gst.region.kegg)])
```

# Pathway Analysis by Methylation Direction

## Identify the direction of significant gene sets

```{r}
sig.GeneSets <- unique(c(gst.region.go$SigGenesInSet[gst.region.go$FDR < 0.05],
                         gst.region.kegg$SigGenesInSet[gst.region.kegg$FDR < 0.05]))
# Split the gene sets into individual genes
split_genes <- function(gene_set) {
 unlist(strsplit(gene_set, split = "[,]"))
}

# Apply the splitting function to all significant gene sets
all_genes <- unique(unlist(lapply(sig.GeneSets, split_genes)))
```

```{r}

# Annotate probes
library(tibble)
top <- rownames_to_column(top, var = "ID")
anno <- as.data.frame(anno)
anno <- rownames_to_column(anno, var = "ID")
top <- merge(top, anno, by = "ID")
```

```{r}
# Split UCSC_RefGene_Name into individual genes
library(tidyr)
top <- top %>%
  mutate(UCSC_RefGene_Name = strsplit(as.character(UCSC_RefGene_Name), split = "[,;]")) %>%
  unnest(UCSC_RefGene_Name)

# Filter for significant gene sets
filteredResults <- top[top$UCSC_RefGene_Name %in% all_genes, ]
```

```{r}
# Summarize methylation status for each significant gene set
methylationSummary <- sapply(all_genes, function(gene) {
 probes <- filteredResults[filteredResults$UCSC_RefGene_Name == gene, ]
 avgLogFC <- mean(probes$logFC)
 return(avgLogFC)
})
```

```{r}
# Determine methylation status
methylationStatus <- ifelse(methylationSummary > 0, "Hypermethylated", "Hypomethylated")

# Combine results into a data frame
results <- data.frame(GeneSet = all_genes, MethylationStatus = methylationStatus)

print(results)
```

## Pathway analysis of genes by methylation direction

```{r, echo = F}
library(clusterProfiler)
library(org.Hs.eg.db)
source("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/dna methyl/gene.list.R")
# create gene list of hypermethylated genes
hyper.gene.list <- gene.list("Hypermethylated")

# create gene list of hypomethylated genes
hypo.gene.list <- gene.list("Hypomethylated")
```

### GO

```{r}
# hypermethylated gene sets
go.hyper.enrich <- enrichGO(gene = hyper.gene.list, OrgDb = org.Hs.eg.db, keyType = "ENTREZID", ont = "ALL")

# hypomethylated gene sets
go.hypo.enrich <- enrichGO(gene = hypo.gene.list, OrgDb = org.Hs.eg.db, keyType = "ENTREZID", ont = "ALL")

# add column specifying methylation direction
go.hyper.enrich <- as.data.frame(go.hyper.enrich)
go.hyper.enrich$MethylationStatus = "Hypermethylated"

go.hypo.enrich <- as.data.frame(go.hypo.enrich)
go.hypo.enrich$MethylationStatus = "Hypomethylated"

# merge data sets
go.enrich <- rbind(go.hyper.enrich, go.hypo.enrich)
```

### KEGG

```{r}
# hypermethylated gene sets
kegg.hyper.enrich <- enrichKEGG(gene = hyper.gene.list, organism = "hsa")

# hypomethylated gene sets
kegg.hypo.enrich <- enrichKEGG(gene = hypo.gene.list, organism = "hsa")

# add column specifying methylation direction
kegg.hyper.enrich <- as.data.frame(kegg.hyper.enrich)
kegg.hyper.enrich$MethylationStatus = "Hypermethylated"

kegg.hypo.enrich <- as.data.frame(kegg.hypo.enrich)
kegg.hypo.enrich$MethylationStatus = "Hypomethylated"

# merge data sets
kegg.enrich <- rbind(kegg.hyper.enrich, kegg.hypo.enrich)
```

Certain interesting gene sets can be pulled in order to determine whether the gene set is significantly hyper or hypomethylated.

```{r}
kegg.enrich[kegg.enrich$ID == "hsa04740", 15]
```

# Export Results

```{r}
write.csv(kegg.enrich, "kegg_SMC_enrichment_results.csv")
write.csv(go.enrich, "go_SMC_enrichment_results.csv")

# create csv of gene list
capture.output(compact(hyper.gene.list), file = "SMC_hypermethyl_genes.txt")
capture.output(compact(hypo.gene.list), file = "SMC_hypomethyl_genes.txt")
```
