---
title: "Longitudinal Analysis"
author: "Kennedy Zinn"
date: "2025-07-21"
output: html_document
---


```{r}
suppressPackageStartupMessages(library(minfi))
suppressPackageStartupMessages(library(limma))
suppressPackageStartupMessages(library(DMRcate))
suppressPackageStartupMessages(library(maxprobes))
suppressPackageStartupMessages(library(readxl))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(missMethyl))
suppressPackageStartupMessages(library(MethylToSNP))
```

```{r}
library(lme4)
```

# Load data
```{r}
# load the idat files
rgSet <- read.metharray.exp("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/dna methyl/within groups/All Phases", force = T)
```

```{r}
# load sample sheet
sample_sheet <- read_excel("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/dna methyl/within groups/All Phases/sample_sheet_ALL.xlsx")
```

# Pre-processing

```{r}
source("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/dna methyl/process_methyl_microarray.R")
gmSet <- process_methyl_microarray(rgSet)
```

Beta values, which range between [0,1], represent the proportion of methylated DNA copies at a specific CpG site compared to all of the DNA copies at that site. Beta values for the ith CpG site are mathematically defined as,

Beta(i) = max(yi(methyl),0) / [max(yi(unmethyl), 0) + max(yi(methyl), 0) + alpha],

where yi methyl and unmethyl are the intensities measured by the ith methylated and unmethylated probes, respectively. 0 ensures that negative signals that take place after background adjustment are not considered. Alpha, a constant (usually 100), serves as an offset to regularize beta when both methylated and unmethylated instensities are low. Beta values are intuitive, and hence useful for visual presentations of the reads. (Du et al., 2010)

```{r, echo = FALSE}
# Compute beta values
beta_values <- getBeta(gmSet)
colnames(beta_values) <- sample_sheet$`Sample Name`
```

M values are log transformed beta values. Negative M values indicate higher levels of unmethylated DNA copies than methylated DNA copies. Postive M values indicate higher levels of methylated DNA copies than unmethylated. M values close to zero indicate similar intensities between methylated and unmethylated probes. M values are mathematically expressed as,

M(i) = log2([max(yi(methyl))+alpha]/[max(yi(unmethyl))+alpha])

where alpha is a constant to offset dramatic changes resulting from small estimation errors. M values can be expressed in terms of beta values,

M(i) = log2(Beta(i)/[1 - Beta(i)])

and therefore, beta values can be expressed in terms of M values,

Beta(i) = 2\^M(i) / [2\^M(i) + 1]

M values are more statistically sound than beta values.

```{r}
# Compute M values
Mval <- getM(gmSet)
dim(Mval)
```

Most variation seems explained by phase (PC1) and ID (PC2). However, proportion of variation is low ((lambda1+lambda2)/total var = 35%)

```{r}
# PCA of M values
par(mfrow=c(1,1), cex =.5)
plotMDS(Mval, labels=sample_sheet$Status, col=as.integer(factor(sample_sheet$Phase)))
legend("top",legend=c("IPSC", "NCC", "SMC"), col=as.integer(factor(sample_sheet$Phase)))
```

PC1 and PC2 by disease status

```{r}
par(mfrow=c(1,1), cex =.5)
plotMDS(Mval, labels=sample_sheet$`Sentrix ID`, col=as.integer(factor(sample_sheet$Status)))
legend("top",legend=c("Control", "Patient"), col=1:2)
```

PC3 and PC4

```{r}
par(mfrow=c(1,1), cex =.5)
plotMDS(Mval, labels=sample_sheet$Phase, col=as.integer(factor(sample_sheet$Status)), dim.plot = 3:4)
legend("top",legend=c("Control","Patient"),col=1:2)
```

# Differential Methylation Analysis

```{r}
# match column names of Mval to row names of sample sheet
sample_sheet <- as.data.frame(sample_sheet)
rownames(sample_sheet) <- paste(sample_sheet$`Sentrix ID`, sample_sheet$`Sentrix Position`, sep = "_")

# reorder sample sheet to match row names to column names of Mval
sample_sheet <- sample_sheet[colnames(Mval),]

# regular interaction model
disease <- factor(sample_sheet$Status)
phase <- factor(sample_sheet$Phase)
sentrixID <- factor(sample_sheet$`Sentrix ID`)
gender <- factor(sample_sheet$Gender)

design <- model.matrix(~gender+sentrixID+disease*phase)
fit <- lmFit(Mval, design)
fit <- eBayes(fit)
top <- topTable(fit, coef=8:9, number = "Inf")
top
```
```{r}
# repeat analysis but with contrasts to delineate changes between phases
design <- model.matrix(~0+gender+sentrixID+disease*phase)
colnames(design) <- make.names(colnames(design))
cont.dif <- makeContrasts(
    ncc_to_smc=(diseasepatient.phaseSMC-diseasepatient.phaseNCC),
    ipsc_to_smc = diseasepatient.phaseSMC,
    ipsc_to_ncc = diseasepatient.phaseNCC,
  levels = design
)
fit2 <- contrasts.fit(fit, cont.dif)
fit2 <- eBayes(fit2)
top <- topTable(fit2, number = "Inf")
top
```

## Kegg Region Analysis
```{r}
region <- function(coef){
  myAnnotation <- cpg.annotate(datatype = "array", object = Mval, what = "M",
                             arraytype = "EPICv1",
                             analysis.type = "differential", design = design, contrasts = T, cont.matrix = cont.dif, coef = coef)
  if (sum(myAnnotation@ranges@elementMetadata@listData$is.sig)==0) {
    return(NULL)
  } else {
    DMRs <- suppressMessages(dmrcate(myAnnotation))
  results.ranges <- extractRanges(DMRs, genome="hg19")
    return(results.ranges)
  }
}

# cpg.annotate() cannot test more than one coefficient at once
# ipsc - NCC
ncc.ranges <- region("ipsc_to_ncc")

# ipsc - SMC
smc.ranges <- region("ipsc_to_smc")

# ncc to smc
ncc_smc.ranges <- region("ncc_to_smc")
```

```{r}
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)

# testing Yan's gene
smc.region <- gsaregion(smc.ranges, all.cpg = rownames(top), collection="6383", array.type="EPIC", anno=anno)

topGSA(smc.region)
```
```{r}
ncc_to_smc.region <- gsaregion(ncc_smc.ranges, all.cpg = rownames(top), collection = "6383", array.type = "EPIC", anno=anno)
topGSA(ncc_to_smc.region)
```


```{r}
# overall gene set testing
go_region <- function(granges, collection){
  set <- goregion(granges, all.cpg = rownames(top), collection = collection, array.type="EPIC", anno=anno, sig.genes = T)
  return(set)
}

smc.set.kegg <- go_region(smc.ranges, "KEGG")
smc.set.go <- go_region(smc.ranges, "GO")
ncc_smc.kegg <- go_region(ncc_smc.ranges, "KEGG")
ncc_smc.go <- go_region(ncc_smc.ranges, "GO")
```
```{r}
# identify significant genes
unique(c(smc.set.go$SigGenesInSet,
         smc.set.kegg$SigGenesInSet,
         ncc_smc.go$SigGenesInSet,
         ncc_smc.kegg$SigGenesInSet))
```


```{r}
# annotate probes
library(tibble)
library(tidyr)
anno.df <- as.data.frame(anno)
anno.df <- rownames_to_column(anno.df, var = "ID")

annotate_probes <- function(region_results, coef){
  top <- topTable(fit2, number = "Inf", coef = coef) %>% 
    rownames_to_column(var = "ID")
  top <- merge(top, anno.df, by = "ID") %>% # split gene list into individual genes
    dplyr::mutate(UCSC_RefGene_Name = strsplit(as.character(UCSC_RefGene_Name), split = "[,;]")) %>%
    unnest(UCSC_RefGene_Name)
  return(top)
}

# separate by genes of interest
top.otx1 <- top[top$UCSC_RefGene_Name=="OTX1",]
top.sdc2 <- top[top$UCSC_RefGene_Name=="SDC2",]
top.sdk1 <- top[top$UCSC_RefGene_Name=="SDK1",]
top.bloc1s2

# determine overall methylation status of gene
mean(top.otx1$logFC) #hypermethylated
mean(top.sdc2$logFC) #hypomethylated, but not noted as a sig gene in region analyses
mean(top.sdk1$logFC) #hypomethylated
```

# Visualize profiles for each sig gene
```{r}
# load rgset for each phase
# process
# obtain topTable
# plot avg logFC per gene x group x phase
```

